File: 2015/day23.pl
#!/usr/bin/env perl 
#===============================================================================
#
#         FILE: day23.pl
#
#        USAGE: ./day23.pl  
#
#  DESCRIPTION: Day 23 - A really crappy computer for a present.
#
#      OPTIONS: ---
# REQUIREMENTS: ---
#         BUGS: ---
#        NOTES: ---
#       AUTHOR: Mark T. Johnson (mtj), markj@matzsoft.com
# ORGANIZATION: MATZ Software & Consulting
#      VERSION: 1.0
#      CREATED: 12/23/15 08:57:28
#     REVISION: ---
#===============================================================================

use strict;
use warnings;
use utf8;
use 5.010;

use Math::BigInt;


my $pc = 0;
my %registers = ( a => Math::BigInt->new( 0 ), b => Math::BigInt->new( 0 ) );
my @memory;
my %opcodes = (
    hlf => { register => 1 },
    tpl => { register => 1 },
    inc => { register => 1 },
    jmp => { offset => 1 },
    jie => { register => 1, offset => 1 },
    jio => { register => 1, offset => 1 },
);


########################################
# VSA - Very Stupid Assembler
########################################

while ( <DATA> ) {
    chomp;
    s/,//g;

    my ( $opcode, $operand1, $operand2 ) = split;

    exists $opcodes{$opcode} or die "$opcode invalid opcode on line $.\n";

    if ( exists $opcodes{$opcode}{register} ) {
        if ( exists $opcodes{$opcode}{offset} ) {
            push @memory, { opcode => $opcode, register => $operand1, offset => $operand2 };
        } else {
            push @memory, { opcode => $opcode, register => $operand1 };
        }

    } elsif ( exists $opcodes{$opcode}{offset} ) {
        push @memory, { opcode => $opcode, offset => $operand1 };

    } else {
        die "Bad opcode table for $opcode\n";
    }
}


########################################
# Now produce the answers
########################################

vmi( 0, 0, 0 );
vmi( 0, 1, 0 );


#######################################
# VMI - Very Moronic Interpreter
########################################

sub vmi {
    $pc = shift;
    $registers{a} = Math::BigInt->new( shift );
    $registers{b} = Math::BigInt->new( shift );

    while ( 0 <= $pc and $pc < @memory ) {
        my $opcode   = $memory[$pc]{opcode};
        my $register = $memory[$pc]{register};
        my $offset   = $memory[$pc]{offset};

        given ( $opcode ) {
            when ( 'hlf' ) {
                $registers{$register}->bdiv( 2 );
            }

            when ( 'tpl' ) {
                $registers{$register}->bmul( 3 );
            }

            when ( 'inc' ) {
                $registers{$register}->badd( 1 );
            }

            when ( 'jmp' ) {
                $pc += $offset - 1;
            }

            when ( 'jie' ) {
                if ( $registers{$register}->is_even ) {
                    $pc += $offset - 1;
                }
            }

            when ( 'jio' ) {
                if ( $registers{$register}->is_one ) {
                    $pc += $offset - 1;
                }
            }

            default {
                die "$opcode - invald opcode at address $pc\n";
            }
        }

        $pc++;
    }

    say "PC = $pc, a = $registers{a}, b = $registers{b}";
}



__END__
jio a, +22
inc a
tpl a
tpl a
tpl a
inc a
tpl a
inc a
tpl a
inc a
inc a
tpl a
inc a
inc a
tpl a
inc a
inc a
tpl a
inc a
inc a
tpl a
jmp +19
tpl a
tpl a
tpl a
tpl a
inc a
inc a
tpl a
inc a
tpl a
inc a
inc a
tpl a
inc a
inc a
tpl a
inc a
tpl a
tpl a
jio a, +8
inc b
jie a, +4
tpl a
inc a
jmp +2
hlf a
jmp -7

File: 2015/day23.swift
//
//         FILE: main.swift
//  DESCRIPTION: day23 - Opening the Turing Lock
//        NOTES: ---
//       AUTHOR: Mark T. Johnson, markj@matzsoft.com
//    COPYRIGHT: © 2021 MATZ Software & Consulting. All rights reserved.
//      VERSION: 1.0
//      CREATED: 07/17/21 15:03:20
//

import Foundation

class Computer {
    struct Instruction {
        enum Opcode: String { case hlf, tpl, inc, jmp, jie, jio }
        
        let opcode: Opcode
        let register: String
        let offset: Int
        
        init( line: String ) throws {
            let words = line.split( whereSeparator: { " ,+".contains( $0 ) } )
            
            guard let opcode = Opcode( rawValue: String( words[0] ) ) else {
                throw RuntimeError( "Invalid opcode \(words[0])." )
            }
            
            self.opcode = opcode
            switch opcode {
            case .hlf:
                register = String( words[1] )
                offset = 0
            case .tpl:
                register = String( words[1] )
                offset = 0
            case .inc:
                register = String( words[1] )
                offset = 0
            case .jmp:
                register = ""
                offset = Int( words[1] )!
            case .jie:
                register = String( words[1] )
                offset = Int( words[2] )!
            case .jio:
                register = String( words[1] )
                offset = Int( words[2] )!
            }
        }
    }
    
    let memory: [Instruction]
    var pc = 0
    var registers = [ "a": 0, "b": 0 ]
    
    init( lines: [String] ) throws {
        memory = try lines.map { try Instruction( line: $0 ) }
    }
    
    func run() -> Void {
        while 0 <= pc && pc < memory.count {
            let instruction = memory[pc]
            
            switch instruction.opcode {
            case .hlf:
                registers[ instruction.register, default: 0 ] /= 2
            case .tpl:
                registers[ instruction.register, default: 0 ] *= 3
            case .inc:
                registers[ instruction.register, default: 0 ] += 1
            case .jmp:
                pc += instruction.offset - 1
            case .jie:
                if registers[ instruction.register, default: 0 ] % 2 == 0 { pc += instruction.offset - 1 }
            case .jio:
                if registers[ instruction.register, default: 0 ] == 1 { pc += instruction.offset - 1 }
            }
            
            pc += 1
        }
    }
}


func parse( input: AOCinput ) -> Computer {
    return try! Computer( lines: input.lines )
}


func part1( input: AOCinput ) -> String {
    let computer = parse( input: input )
    
    computer.run()
    return "\( computer.registers["b"]! )"
}


func part2( input: AOCinput ) -> String {
    let computer = parse( input: input )
    
    computer.registers["a"] = 1
    computer.run()
    return "\( computer.registers["b"]! )"
}


try runTests( part1: part1 )
try runTests( part2: part2 )
try solve( part1: part1 )
try solve( part2: part2 )
